src/stdio_mgr/types.py

```diff
@@ -0,0 +1,179 @@
+            return None
+
+        @classmethod
+        def __subclasshook__(cls, subclass):
+            """Check whether subclass is considered a subclass of this ABC."""
+            if cls is AbstractContextManager:
+                return _collections_abc._check_methods(
+                    subclass, "__enter__", "__exit__"
+                )
+            return NotImplemented
+
+
+class TupleContextManager(tuple, AbstractContextManager):
+    """Base for context managers that are also a tuple."""
+
+    # This is needed to establish a workable MRO.
+
+
+class StdioTuple(TupleContextManager):
+    """Tuple context manager of stdin, stdout and stderr stream-like objects."""
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three streams."""
+        items = list(iterable)
+        assert len(items) == 3, "iterable must be three items"  # noqa: S101
+
+        return super(StdioTuple, cls).__new__(cls, items)
+
+    @property
+    def stdin(self):
+        """Return capturing stdin stream."""
+        return self[0]
+
+    @property
+    def stdout(self):
+        """Return capturing stdout stream."""
+        return self[1]
+
+    @property
+    def stderr(self):
+        """Return capturing stderr stream."""
+        return self[2]
+
+    def _map_method(self, method: str):
+        """Perform method on all streams."""
+        for item in self:
+            yield item.__dict__[method](item)
+
+    def _map(self, op):
+        """Return generator for performing op on all streams."""
+        if isinstance(op, str):
+            return self._map_method(op)
+
+        return map(op, self)
+
+    def suppress_map(self, ex, op):
+        """Return generator for performing op on all streams, suppressing ex."""
+        for item in self:
+            with suppress(ex):
+                op_inner = item.__dict__[op] if isinstance(op, str) else op
+                yield op_inner(item)
+
+    def _all(self, op):
+        """Perform op on all streams, returning True when all were successful."""
+        return all(self._map(op))
+
+    def suppress_all(self, ex, op):
+        """Perform op on all streams, suppressing ex."""
+        return all(self.suppress_map(ex, op))
+
+    def close(self):
+        """Close all streams."""
+        return self._all("close")
+
+
+class TextIOTuple(StdioTuple):
+    """Tuple context manager of stdin, stdout and stderr TextIOBase objects."""
+
+    # pytest and colorama inject objects into sys.std* that are not real TextIOBase
+    # and fail the assertion of this class
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three TextIOBase streams."""
+        self = super(StdioTuple, cls).__new__(cls, iterable)
+        assert self._all(  # noqa: S101
+            lambda item: isinstance(item, TextIOBase)
+        ), "iterable must contain only TextIOBase"
+        return self
+
+
+class _MultiCloseContextManager(StdioTuple):
```

Comment:
This should probably have `TupleContextManager` as its base, as it isnt limited to 3 streams.

but keeping the order simple atm is helpful.

I'd like to make sure that the MROs for combinations needed for https://github.com/bskinn/stdio-mgr/pull/64 and https://github.com/bskinn/stdio-mgr/pull/78 also work, so need more tests if this doesnt have `StdioTuple` as its direct superclass

Reply:
Removed from this PR.